src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -269,14 +283,32 @@ void ClusterTaskManager::TasksUnblocked(const std::vector<TaskID> ready_ids) {
       waiting_tasks_.erase(it);
     }
   }
+  ScheduleAndDispatchTasks();
 }
 
-void ClusterTaskManager::HandleTaskFinished(std::shared_ptr<WorkerInterface> worker) {
-  cluster_resource_scheduler_->FreeLocalTaskResources(worker->GetAllocatedInstances());
-  worker->ClearAllocatedInstances();
-  cluster_resource_scheduler_->FreeLocalTaskResources(
-      worker->GetLifetimeAllocatedInstances());
-  worker->ClearLifetimeAllocatedInstances();
+void ClusterTaskManager::HandleTaskFinished(std::shared_ptr<WorkerInterface> worker,
+                                            Task *finished_task) {
+  if (finished_task) {
+    *finished_task = worker->GetAssignedTask();
+  }
+
+  if (worker->GetAllocatedInstances() != nullptr) {
+    cluster_resource_scheduler_->FreeLocalTaskResources(worker->GetAllocatedInstances());
+    worker->ClearAllocatedInstances();
+    cluster_resource_scheduler_->FreeLocalTaskResources(
+        worker->GetLifetimeAllocatedInstances());
+    worker->ClearLifetimeAllocatedInstances();
+  }
+}
+
+void ClusterTaskManager::ReturnWorkerResources(std::shared_ptr<WorkerInterface> worker) {
```

Comment:
Some methods (e.g. `ReturnWorkerResources` and `FreeLocalTaskResources`) in ClusterTaskManager only deal with resources and only call methods in the `ClusterResourceScheduler`.  They have nothing to do with "tasks". 
So I think they should better be put in the `ClusterResourceScheduler`.

Reply:
I dont think so, both `ReturnWorkerResources` and `FreeLocalTaskResources` invoke the interface of `worker`(e.g. `worker->ClearLifetimeAllocatedInstances()`). The `ClusterResourceScheduler` is now only deal with the resources, I think it should not see and operate the `worker`.

Reply:
Hmm I think I agree with @raulchen here, though. If those methods don't actually touch any `ClusterTaskManager` state, it seems a bit strange to implement the methods there.

Reply:
- I've tried to abstract interfaces to `clusters_resource_scheduler_`. However, it is difficult to abstract because the new schedule relies on `TaskResourceInstances`, while the old schedule relies on `ResourceSet`.
- In addition, the old schedule also relies on `local_ queues_` which does not exist in the new scheduler.
cc @raulchen @stephanie-wang